compiler/rustc_hir_analysis/src/collect/predicates_of.rs

```diff
@@ -850,3 +861,170 @@ impl<'tcx> ItemCtxt<'tcx> {
         bounds.clauses(self.tcx).collect()
     }
 }
+
+/// Compute the conditions that need to hold for a conditionally-const item to be const.
+/// That is, compute the set of `~const` where clauses for a given item.
+///
+/// This query also computes the `~const` where clauses for associated types, which are
+/// not "const", but which have item bounds which may be `~const`. These must hold for
+/// the `~const` item bound to hold.
+pub(super) fn const_conditions<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: LocalDefId,
+) -> ty::ConstConditions<'tcx> {
+    // This logic is spaghetti, and should be cleaned up. The current methods that are
+    // defined to deal with constness are very unintuitive.
+    if tcx.is_const_fn_raw(def_id.to_def_id()) {
+        // Ok, const fn or method in const trait.
+    } else {
+        match tcx.def_kind(def_id) {
+            DefKind::Trait => {
+                if !tcx.is_const_trait(def_id.to_def_id()) {
+                    return Default::default();
+                }
+            }
+            DefKind::Impl { .. } => {
+                if !tcx
+                    .trait_id_of_impl(def_id.to_def_id())
+                    .is_some_and(|def_id| tcx.is_const_trait(def_id))
+                {
+                    return Default::default();
+                }
+            }
+            DefKind::AssocTy | DefKind::AssocFn => {
+                let parent_def_id = tcx.local_parent(def_id).to_def_id();
+                match tcx.associated_item(def_id).container {
+                    ty::AssocItemContainer::TraitContainer => {
+                        if !tcx.is_const_trait(parent_def_id) {
+                            return Default::default();
+                        }
+                    }
+                    ty::AssocItemContainer::ImplContainer => {
+                        if !tcx
+                            .trait_id_of_impl(parent_def_id)
+                            .is_some_and(|def_id| tcx.is_const_trait(def_id))
+                        {
+                            return Default::default();
+                        }
+                    }
+                }
+            }
+            _ => return Default::default(),
+        }
+    }
+
+    let (generics, is_trait, has_parent) = match tcx.hir_node_by_def_id(def_id) {
```

Comment:
rename `is_trait` to `trait_id_and_supertraits` or something else that is clearer